### ISSUE TITLE
Can't set the border properties of the NumberBox.

### ISSUE THREAD

**Issue by hipo760** [microsoft/microsoft-ui-xaml#1925]
**Describe the bug**

Border properties like BorderBrush or BorderThickness can't change from xaml.

**Steps to reproduce the bug**

`<ComboBox Header="ComboxBorder" BorderThickness="0.5" BorderBrush="DarkGray" />`
`<controls:NumberBox Header="ComboxBorder" BorderThickness="0.5"  BorderBrush="DarkGray"/>`

**Expected behavior**

Border properties should work like other controls (ComboBox, TextBox,...)

**Screenshots**

![image](https://user-images.githubusercontent.com/55172914/73626431-d49de880-4683-11ea-8d06-a3640396be90.png)

**Version Info**

NuGet package version: 

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| November 2019 Update (18363)  | Yes   |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 | Yes   |
| Mobile                   |    |
| Xbox                      |    |
| Surface Hub          |     |
| IoT                         |     |

**Additional context**

**Comment by ranjeshj** [comment]
@SavoySchuler @teaP Should we expose these TextBox properties onto NumberBox through a template bind ? Are there any other TextBox properties that would fall under the same category ?

**Comment by SavoySchuler** [comment]
@ranjeshj - thanks for the chat, that's exactly what we should do. @teaP is there anything you need from my end to expose these TextBox properties onto NumberBox through a template bind?

**Comment by Felix-Dev** [comment]
@teaP Are you currently working on this (or soon) or is this issue up for grabs for the community?

**Comment by ranjeshj** [comment]
@Felix-Dev, it is up for grabs now :)

**Comment by Felix-Dev** [comment]
In that case, I would like to tackle this one.

**Comment by Felix-Dev** [comment]
To clarify: In SpinButtonPlacementMode.Inline, if we apply properties like BorderThickness and BorderBrush, is the desired look the following?

```xml
<controls:NumberBox
    BorderBrush="Orange"
    BorderThickness="4"/>
```

![image](https://user-images.githubusercontent.com/1398851/80014752-08ddb400-84d1-11ea-9cf7-37602588a012.png)

**Comment by mdtauk** [comment]
> To clarify: In SpinButtonPlacementMode.Inline, if we apply properties like BorderThickness and BorderBrush, is the desired look the following?
> 
> ![image](https://user-images.githubusercontent.com/1398851/80014752-08ddb400-84d1-11ea-9cf7-37602588a012.png)

I would say so yes

**Comment by hipo760** [comment]
> To clarify: In SpinButtonPlacementMode.Inline, if we apply properties like BorderThickness and BorderBrush, is the desired look the following?
> 
> ```
> <controls:NumberBox
>     BorderBrush="Orange"
>     BorderThickness="4"/>
> ```
> 
> ![image](https://user-images.githubusercontent.com/1398851/80014752-08ddb400-84d1-11ea-9cf7-37602588a012.png)

I can't image if the BorderThickness=0, what will spin button looks like?

**Comment by Felix-Dev** [comment]
@hipo760 
Setting the BorderThickness to 0 for the NumberBox would result in the following look:
![image](https://user-images.githubusercontent.com/1398851/80400212-c8a87800-88ba-11ea-8210-d1869b11a657.png)

and mouse hovering over the increase button:
![image](https://user-images.githubusercontent.com/1398851/80400296-dfe76580-88ba-11ea-8a35-97f135fa361a.png)

**Comment by rwx777** [comment]
> @hipo760
> Setting the BorderThickness to 0 for the NumberBox would result in the following look:
> ![image](https://user-images.githubusercontent.com/1398851/80400212-c8a87800-88ba-11ea-8210-d1869b11a657.png)
> 
> and mouse hovering over the increase button:
> ![image](https://user-images.githubusercontent.com/1398851/80400296-dfe76580-88ba-11ea-8a35-97f135fa361a.png)

That looks so great! Thank you.

**Comment by hipo760** [comment]
> @hipo760
> Setting the BorderThickness to 0 for the NumberBox would result in the following look:
> ![image](https://user-images.githubusercontent.com/1398851/80400212-c8a87800-88ba-11ea-8210-d1869b11a657.png)
> 
> and mouse hovering over the increase button:
> ![image](https://user-images.githubusercontent.com/1398851/80400296-dfe76580-88ba-11ea-8a35-97f135fa361a.png)

Looks so great, thank you!!

**Comment by Felix-Dev** [comment]
@ranjeshj Currently, we expose the border thickness of the NumberBox Inline SpinButtons as a resource:
```
<Thickness x:Key="NumberBoxSpinButtonBorderThickness">0,1,1,1</Thickness>
```

Note that we don't set the left border side here. This way, the border between the two spin buttons has the same thickness as the other border sides (e.g. top, bottom) as can be seen in the images above. 
In order to make the SpinButtons border settable via a BorderThickness TemplateBinding (so that NumberBox.BorderThickness affects the whole NumberBox and not just its internal TextBox) I could add a new API similar to the [CornerRadiusFilter API](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Common/CornerRadiusFilterConverter.cpp. The API would enable us to specify a *mask* to apply to a `Thickness` value in order to ignore the thickness value for specific sides. With the CornerRadiusFilter API we specify which corner radius values to keep using the `Filter` property. The Thickness masking API would operate the other way: Specifying the border side we want to ignore (set to 0).

An example use case for the NumberBox could look like this:
```xml
<!-- The Thickness masking resource to use -->
<primitives:ThicknessMaskConverter x:Key="LeftThicknessMaskConverter" Mask="Left"/>
```

```xml
<!-- Use the above Thickness masking resource for NumberBox' SpinButtons in Inline mode -->
<RepeatButton x:Name="DownSpinButton"
    BorderThickness="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=BorderThickness, Converter={StaticResource LeftThicknessMaskConverter}}"
    Style="{StaticResource NumberBoxSpinButtonStyle}"/>
```

Thoughts?

**Comment by ranjeshj** [comment]
@Felix-Dev That sounds like a reasonable approach. This would be new API though, so adding @SavoySchuler and @MikeHillberg.

**Comment by SavoySchuler** [comment]
Just assigned myself so that I can take a look at this.

**Comment by Felix-Dev** [comment]
@SavoySchuler I have finalized my API proposal now (including a test implementation in a local branch of mine).

To recap, the NumberBox uses three borders to achieve its visual look in all the different visual states: A border around its input field (the TextBox and its border) and around each of its SpinButtons when in SpinButtonPlacementMode::Inline. See the following image of a NumberBox in rest mode:
![image](https://user-images.githubusercontent.com/1398851/86408623-9c90e500-bcb7-11ea-95e2-92d39e5a3e4a.png)
To achieve the uniform border thickness used here, the border thickness values of the spin buttons are of the form `0,X,X,X` (so the border's left side thickness is hard-coded to 0).

Currently, setting the `NumberBox.BorderBrush` API has no effect on the NumberBox which @hipo760 lists in their opening post. If one wishes to easily customize the border thickness of  aNumberBox, lightweight styling has to be used instead for now. To add support for the `NumberBox.BorderThickness` API and make it work as intended, we need to *filter* the template bound *BorderThickness* for the spin buttons accordingly. In other words, we need to set the `Thickness.Left` property of a given `Thickness` value to 0 (zero). See the following GIF which shows what will happen to the NumberBox's border if we template bind to the `NumberBox.BorderThickness` API without filtering the thickness value as described above:
```xml
<RepeatButton x:Name="UpSpinButton"
    BorderThickness="{TemplateBinding BorderThickness}"/>
```
![numberbox-borderthickness-withoutAPI](https://user-images.githubusercontent.com/1398851/86410663-78cf9e00-bcbb-11ea-9b6e-1b9b1ddc2eed.gif)

As a comparison, here is how the `NumberBox.BorderThickness` API will be applied *with* correct filtering of the template bound BorderThickness:
![numberbox-borderthickness-withAPI](https://user-images.githubusercontent.com/1398851/86411121-6f930100-bcbc-11ea-9524-61b235241b4d.gif)

To achieve this, I propose adding a new API similar to the existing [CornerRadiusFilterConverter API](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.primitives.cornerradiusfilterconverter?view=winui-2.4). See below for an API proposal and an API example.

### API proposal
```cpp
[flags]
enum ThicknessMaskKinds
{
    None = 0,
    Top = 1,
    Right = 2,
    Bottom = 4,
    Left = 8
};

runtimeclass ThicknessMaskConverter : Windows.UI.Xaml.DependencyObject, Windows.UI.Xaml.Data.IValueConverter
{
    ThicknessMaskConverter();

    [MUX_DEFAULT_VALUE("winrt::ThicknessMaskKinds::None")]
    ThicknessMaskKinds Mask{ get; set; };

    winrt::IInspectable ThicknessMaskConverter::Convert(
        winrt::IInspectable const& value, 
        winrt::TypeName const& targetType, 
        winrt::IInspectable const& parameter, 
        winrt::hstring const& language);

    winrt::IInspectable ThicknessMaskConverter::ConvertBack(
        winrt::IInspectable const& value,
        winrt::TypeName const& targetType,
        winrt::IInspectable const& parameter,
        winrt::hstring const& language);

    static Windows.UI.Xaml.DependencyProperty MaskProperty{ get; };
};
```

### API Notes
| Property | Description |
|----------|--------------|
| Mask | Gets or sets the mask applied to the ThicknessMaskConverter specifying which values of a `Thickness` struct will be set to 0 (zero) and which to keep. Can be a combination of multiple values. |

### API Example
Consuming the proposed API is similar to how you would use the `CornerRadiusFilterConverter` API. See, for example, the following XAML below. Here, we first create a new instance of the `ThicknessMaskConverter` class with our desired `Mask` and then use our converter in a binding. For demo purposes, I created an example `Thickness` resource called *ExampleThickness* in order to use a binding. In the NumberBox ControlTemplate case, we would instead use the much more commonly used TemplatedParent relative binding.
```xml
<Page
    xmlns:primitives="using:Microsoft.UI.Xaml.Controls.Primitives">
    <Page.Resources>
        <primitives:ThicknessMaskConverter x:Key="LeftTopRightThicknessMaskConverter" Mask="Left,Top,Right"/>

        <Thickness x:Key="ExampleThickness">4</Thickness>
    </Page.Resources>

    <Grid x:Name="RootGrid" Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
        <TextBox
            Text="I am a TextBox with just a bottom border"
             BorderBrush="GreenYellow"
            BorderThickness="{Binding Source={StaticResource ExampleThickness}, Converter={StaticResource LeftTopRightThicknessMaskConverter}}" />
    </Grid>
</Page>
```

The XAML above gives us the following TextBox look:
![image](https://user-images.githubusercontent.com/1398851/86414025-17133200-bcc3-11ea-8aac-c9fc9d25c664.png)

As already mentioned above, for a real word example, we would use this API in the NumberBox ControlTemplate like this:
```xml
<RepeatButton x:Name="UpSpinButton"
    BorderThickness="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=BorderThickness, Converter={StaticResource LeftThicknessMaskConverter}}"
    Style="{StaticResource NumberBoxSpinButtonStyle}"/>
```
where *LeftThicknessMaskConverter* is defined as:
```xml
<primitives:ThicknessMaskConverter x:Key="LeftThicknessMaskConverter" Mask="Left"/>
```

### Additional Notes
* First and foremost, I'd like to give a shoutout to @stevenbrix who is always a great help when I have specific implementation questions. Thank you, Steven!

* Secondly, while we can view this API as *filtering* out specific `Thickness` values and throwing away others (by setting them to *0*) I went with the "Mask" terminology for now. The reason being that I can imagine more of than not we are only interested in in ignoring the value of 1-2 thickness values instead of 3-4. Take our NumberBox case here, for example. We filter out the *Top*, *Right* and *Bottom* values of a `Thickness` value and ignore only the `Left` value. If I would have went with the *filter* terminology, here's how my my XAML would have looked like instead:
```xml
<primitives:ThicknessFilterConverter x:Key="LeftTopRightThicknessFilterConverter" Filter="Left,Top,Right"/>

<RepeatButton x:Name="UpSpinButton"
    BorderThickness="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=BorderThickness, Converter={StaticResource LeftTopRightThicknessFilterConverter}}"
    Style="{StaticResource NumberBoxSpinButtonStyle}"/>
```
Compare that to the XAML using the `Mask` terminology:
```xml
<primitives:ThicknessMaskConverter x:Key="LeftThicknessMaskConverter" Mask="Left"/>
```
Obviously, I am open to renaming and restructing the proposed API if the team prefers the name *ThicknessFilterConverter* (or any other name) over my proposed naming.

## Summary
The proposed `ThicknessMaskConverter` API will be a great helper API to achieve proper `NumberBox.BorderThickness` support. More broadly speaking it is to be used in a (Template) Binding to create a new `Thickness` struct from an existing one, setting only some of the fields to 0 while keeping the rest.

**Comment by Felix-Dev** [comment]
@SavoySchuler Any update here?

**Comment by ldeffenb** [comment]
Is this issue just withering on the vine? Or has something that accomplishes this been implemented, but not linked to this issue?  I'm definitely interested in this functionality.

**Comment by chingucoding** [comment]
I think this was fixed with #4469
